client/history/caching_redirector.go

```diff
@@ -0,0 +1,213 @@
+// THE SOFTWARE.
+
+package history
+
+import (
+	"context"
+	"errors"
+	"sync"
+
+	"go.temporal.io/api/serviceerror"
+
+	"go.temporal.io/server/api/historyservice/v1"
+	"go.temporal.io/server/common"
+	"go.temporal.io/server/common/log"
+	"go.temporal.io/server/common/log/tag"
+	"go.temporal.io/server/common/membership"
+	serviceerrors "go.temporal.io/server/common/serviceerror"
+)
+
+type (
+	cacheEntry struct {
+		shardID    int32
+		address    rpcAddress
+		connection clientConnection
+	}
+
+	// A cachingRedirector is a redirector that maintains a cache of shard
+	// owners, and uses that cache instead of querying membership for each
+	// operation. Cache entries are evicted either for shard ownership lost
+	// errors, or for any error that might indicate the history instance
+	// is no longer available, including timeouts.
+	cachingRedirector struct {
+		mu struct {
+			sync.RWMutex
+			cache map[int32]cacheEntry
+		}
+
+		connections            connectionPool
+		historyServiceResolver membership.ServiceResolver
+		logger                 log.Logger
+	}
+)
+
+func newCachingRedirector(
+	connections connectionPool,
+	historyServiceResolver membership.ServiceResolver,
+	logger log.Logger,
+) *cachingRedirector {
+	r := &cachingRedirector{
+		connections:            connections,
+		historyServiceResolver: historyServiceResolver,
+		logger:                 logger,
+	}
+	r.mu.cache = make(map[int32]cacheEntry)
+	return r
+}
+
+func (r *cachingRedirector) clientForShardID(shardID int32) (historyservice.HistoryServiceClient, error) {
+	if err := checkShardID(shardID); err != nil {
+		return nil, err
+	}
+	entry, err := r.getOrCreateEntry(shardID)
+	if err != nil {
+		return nil, err
+	}
+	return entry.connection.historyClient, nil
+}
+
+func (r *cachingRedirector) execute(ctx context.Context, shardID int32, op clientOperation) error {
+	if err := checkShardID(shardID); err != nil {
+		return err
+	}
+	opEntry, err := r.getOrCreateEntry(shardID)
+	if err != nil {
+		return err
+	}
+	return r.redirectLoop(ctx, opEntry, op)
+}
+
+func (r *cachingRedirector) redirectLoop(ctx context.Context, opEntry cacheEntry, op clientOperation) error {
+	for {
+		if err := common.IsValidContext(ctx); err != nil {
+			return err
+		}
+		opErr := op(ctx, opEntry.connection.historyClient)
+		if opErr == nil {
+			return opErr
+		}
+		if maybeHostDownError(opErr) {
+			r.cacheDeleteByAddress(opEntry.address)
+			return opErr
+		}
+		var solErr *serviceerrors.ShardOwnershipLost
+		if !errors.As(opErr, &solErr) {
+			return opErr
+		}
+		var again bool
+		opEntry, again = r.handleSolError(opEntry, solErr)
+		if !again {
+			return opErr
+		}
+	}
+}
+
+func (r *cachingRedirector) getOrCreateEntry(shardID int32) (cacheEntry, error) {
+	r.mu.RLock()
+	entry, ok := r.mu.cache[shardID]
+	r.mu.RUnlock()
+	if ok {
+		return entry, nil
+	}
+
+	r.mu.Lock()
+	defer r.mu.Unlock()
+
+	entry, ok = r.mu.cache[shardID]
+	if ok {
+		return entry, nil
+	}
```

Comment:
This is a neat pattern; is there a name for it? If so, maybe comment it. Otherwise, it might look a little weird that we're duplicating the logic, even though it makes sense. It took me a minute to understand